### ISSUE TITLE
Does not work on Ubuntu 22.04

### ISSUE THREAD

**Issue by fakrep** [Ckath/mergeinputs#3]
I installed gcc and compiled mergeinputs on a freshly installed Ubuntu 22.04 LTS.

If I run following command in terminal "sudo mergeinputs /dev/input/by-id/*-kbd" my keyboard is not working anymore. I have to close the terminal. (Also tried with "sudo mergeinputs /dev/input/by-path/*-kbd")

With Ubuntu 20.04 everything was working

Has anyone an idea how to get merginputs running on Ubuntu 22.04?

**Comment by Ckath** [comment]
couple questions, though I dont have an ubuntu install to test against nor a system where it seems to reproduce not working.

1. what's connected to your system, anything weird that might not like it?
`ls /dev/input/by-id/*-kbd`
2. can you see anything complaining in syslog when you run mergeinput and your keyboard disappears?
`dmesg -w` in another terminal before running mergeinputs
4. does it still occur when you only try a couple? or even just one keyboard device
you might be able to isolate it to a device that causes it to not work
5. do keyboards work at all when you plug them in (regardless of mergeinputs) as the merged input device is added when you run it, maybe your setup somehow cant handle it to begin with

also as a side note, what convinces you its ubuntu? if anything it'd be more likely to be your connected to your de if anything. possibly kernel since it does rely on uinput, but I doubt thats the cause.

**Comment by fakrep** [comment]
Thank you for your help.

So far I tried following:

1. What's connected to your system, anything weird that might not like it?
`sudo ls /dev/input/by-id/*kbd
/dev/input/by-id/usb-0911_2188-event-kbd  /dev/input/by-id/usb-Hantick_USB_Keyboard_HID_Composite_device-event-kbd
`

I use the r-go split keyboard. 
`usb-0911_2188-event-kbd` is the left part of the keyboard
`usb-Hantick_USB_Keyboard_HID_Composite_device-event-kbd` is the right part 

When I run `sudo mergeinputs /dev/input/by-id/usb-0911_2188-event-kbd` the left part of the keyboard is no more recognized. Same applies for the right part.

2. Output of `dmesg -w` when I run `mergeinputs` is:
 `input: merged input as /devices/virtual/input/input59`
the number behind input gets incremented each time I run mergeinputs

The keyboard is fully functional without mergeinputs, but if I type quickly with both parts (e.g. 1919191) the display gets laggy / freezes shortly.

With Ubuntu 20.04 LTS on same machine this behavior was fixed by simply running mergeinputs (Thank you so much btw <3 )

**Comment by Ckath** [comment]
hm, so it did work on an older ubuntu, not suspecting any userspace things capable of ruining such functionality the only thing left to blame would be the kernel in that case. I never had issues with the input devices.
the input device does get created it seeems as noted by your dmesg. really only leaves uinput.. but thats also strange on its own as it'd be crucial to alot of other things.

kinda stumped at this point, would also make debugging useless as on the mergeinput side everything looks alright. I'd be tempted to say a new kernel makes it outright not work.. but I'm on an up to date system where it still works (5.19.1 kernel). the only thing I could suggest at this point is trying to see if its limited to this kernel and if you can get around it with some alternative kernels, though thats not a great experience with ubuntu either.

aditionally, I was going to add some more debug printing to it to help the process along but I notice I already put error message on all cases that'd cause this behaviour like being unable to open the output device or any of the input devices

**Comment by Ckath** [comment]
bit of an overly obvious question, and it shouldnt matter anyway.
but I didnt specifically ask it..
did you recompile on your new ubuntu or use the binary from the old one?

**Comment by Ckath** [comment]
it seems to work on this version for others #4 so maybe its something within your userspace programs. I'm afraid I cant think of anyway to debug it further from the mergeinputs side though as everything comes up fine (not a single error on failing to get the input, or the file, or creating the output)

**Comment by Bastien-mva** [comment]
Hey, I am currently getting the same issue. 

I have one laptop with Ubuntu 22.04 working good, even though I do not run mergeinputs, and a desktop computer where it is not working. 

On my desktop, I am experiencing the same issus as @fakrep . My keyboard works but gets big delays when typing left and right keyboards at the same time, and when I run ``` sudo mergeinputs  /dev/input/by-id/*-kbd```, my keyboard does not work anymore. I have tried with ``` sudo mergeinputs /dev/input/by-path/*-kbd ``` also, without success.

**Comment by joakim-developer** [comment]
Also in the same boat! let's figure this out. will be posting if I find a solution

**Comment by Ckath** [comment]
since this hooks in at uinput level, before any xinput or whatnot drivers can get to it. can you confirm it does or doesnt work in another tty without graphics? just logging into a tty (`ctrl` `alt` `f3` or some other one you dont use).
if this is the case its something within whatever the new ubuntu ships there.
also I never asked but are any of you using wayland? I dont see why it would break on wayland since again its before any other input handling but you never know

**Comment by joakim-developer** [comment]
I've been at it since I made the comment 5 hours ago. broke my OS while trying to get wayland to work on Pop-os. tried again after a system recovery and gave up on it. In my case I had no issues with input lag on fedora that uses wayland by default. and also the manufacturer of R-go split keyboard recommended using it before x11 so that must be the issue. 

Funny thing is that ubuntu 22.04 uses wayland by default also. so @Bastien-mva  I would check if you are using wayland on the desktop by running `echo $XDG_SESSION_TYPE`, if you are using x11 then there should be a cog in the login screen where you can select what you wish to use.

edit: **Writing a guide here for other users since I found the way to enable wayland on ubuntu, mint and pop os**

-  The default display manager for the GNOME desktop environment is GDM3. Therefore, we will edit the /etc/gdm3/custom.conf file to either disable or enable Wayland. open a terminal and use nano or your favorite text editor to open this file with root permission AKA sudo

`sudo nano /etc/gdm3/custom.conf`

- Within this file, look for the line that says `#WaylandEnable=false.` You can uncomment this line and either set it to true or false, depending on whether you want Wayland enabled or not.

`WaylandEnable=true`

- After you have made the desired changes, save this file and exit it. You will need to restart GDM3 or reboot for the changes to take effect.

_You may be lucky_ and have the cog on bottom right show on login screen at first attempt but at least for pop users this is most likely not to show.  due to some other code actually skipping your changes located it gdm3. see: [this link](https://wiki.archlinux.org/title/GDM#Wayland_and_the_proprietary_NVIDIA_driver) for more information,
**To solve this** you need to run
`ln -s /dev/null /etc/udev/rules.d/61-gdm.rules` This command works because rules in /etc/udev/rules.d/ override those in /usr/lib/udev/rules.d/ 

Sorry for spamming the thread but I do really hope some would find it helpfull

**Comment by Bastien-mva** [comment]
Thanks for the reply, 

I am working remotely from my desktop computer so I can't try now the r-go split keyboard and will not be able until monday, sorry. I can tell you that I am using x11 on my desktop computer (it's weird no since you are telling me that it should be wayland ?). 

I have wayland on my laptop computer (where the keyboard is working just fine). 

I will try to install wayland on my laptop and see if it works.

Thanks !

**Comment by Bastien-mva** [comment]
Hello, 

I tried to enable wayland but it does not seem to be working. I have nvidia drivers, blocking wayland. I am trying to switch to wayland anyway. 

It does seem to be x11 the issue.

**Comment by Bastien-mva** [comment]
I just managed to enable wayland and it is now working just fine, but mergeinputs did not change anything.

**Comment by Ckath** [comment]
> It does seem to be x11 the issue.

I use x11, the entire point of this thing was it sits before any userspace drivers. just regular x11 systems are completely fine with it.
I dont know what ubuntu started doing in this version. if anything my question if people were using wayland was because I expected it to not work there.

**Comment by joakim-developer** [comment]
Intersting. What OS are you on? I think this is a debian issue so ubuntu and pop-os. It's hard to know about other distros since their on wayland

**Comment by Ckath** [comment]
its just arch with a simple wm, on x11 since forever.
I'd be surprised if its debian, as mentioned before the only way I can see correlations if it'd be something in the kernel. but it cant be the kernel version as I also usually update.
also happen to have a terrible keyboard situation hence this exists as personal slapped together solution
![image](https://user-images.githubusercontent.com/15038414/190709682-c061e809-657e-43b9-a6d9-3a28a5bdb3f1.png)

bit of a re-cap but I just like complaining how non of this makes sense:
I really thought this would forever be fine as its way before any userspace drivers, and while personal experiences confirm now ubuntu seems to be doing _something_ to break it. it not spitting out errors means it can open the devices fine, people say they exist, it even manages to correctly make its own input device. still completely at loss. the people switching to wayland as 'fix' is even more an insult to injury

**Comment by Ckath** [comment]
I'm planning on rolling out debian with xfce for some people in the near future, I'll try to test it there to see if I find any issues similar to this

**Comment by joakim-developer** [comment]
Thanks for going that extra mile! Yeah I see your point, and taking the time to go down the "rabbit hole" to figure it out is 🥳

**Comment by gcalus** [comment]
Hi, i have the same issue (r-go split keyboard lags, ubuntu 22.04, x11, switching to wayland solves it but MS teams doesn't work properly on Wayland)

I've tested mergeinputs and found out that after typing 

`sudo mergeinputs /dev/input/by-id/usb-0911_2188-event-kbd`

the console freezes, keyboard freezes (left part) and command `tail -f /var/log/syslog` outputs:

```
Sep 18 21:42:59 PL-MCALUS kernel: [ 2063.363890] input: merged input as /devices/virtual/input/input45
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (II) config/udev: Adding input device merged input (/dev/input/event264)
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (**) merged input: Applying InputClass "libinput pointer catchall"
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (**) merged input: Applying InputClass "libinput keyboard catchall"
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (II) Using input driver 'libinput' for 'merged input'
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (EE) systemd-logind: failed to take device /dev/input/event264: No such device
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (**) merged input: always reports core events
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (**) Option "Device" "/dev/input/event264"
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (EE) xf86OpenSerial: Cannot open device /dev/input/event264
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: #011Permission denied.
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (II) event264: opening input device '/dev/input/event264' failed (Permission denied).
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (II) event264 - failed to create input device '/dev/input/event264'.
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (EE) libinput: merged input: Failed to create a device for /dev/input/event264
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (EE) PreInit returned 2 for "merged input"
Sep 18 21:42:59 PL-MCALUS /usr/libexec/gdm-x-session[16580]: (II) UnloadModule: "libinput"
```

logs on wayland:

```
Sep 18 22:05:16 PL-MCALUS kernel: [ 3400.319036] input: merged input as /devices/virtual/input/input47
Sep 18 22:05:16 PL-MCALUS gnome-shell[26668]: Could not open device /dev/input/event264: GDBus.Error:System.Error.ENODEV: No such device
```

Strange enough that logs say `Adding input device merged input (/dev/input/event264)` and just after that `failed to take device /dev/input/event264: No such device` - maybe ubuntu 22.04 is somehow getting in the way of creating the input?

---

Could be related: pressing a key from different keyboard than previous (f-j-f-j, or even the same key but between notebook and external keyboard) gives this logs:

```
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 31 with keysym 31 (keycode a).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 34 with keysym 34 (keycode d).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 35 with keysym 35 (keycode e).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 36 with keysym 36 (keycode f).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 37 with keysym 37 (keycode 10).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 38 with keysym 38 (keycode 11).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 39 with keysym 39 (keycode 12).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 32 with keysym 32 (keycode b).
Sep 18 21:44:19 PL-MCALUS gnome-shell[17104]: Window manager warning: Overwriting existing binding of keysym 33 with keysym 33 (keycode c).
```

But i think it's not the issue, because uninstalling the package gnome-shell-extension-ubuntu-dock makes warnings go away without resolving the keyboard lag issue.

**Comment by Ckath** [comment]
ah this is slightly more insightful, I might have an idea whats going on. in the lands of less managed desktop setups without much management software those /dev/input/ symlinks are all owned by root. so it doesnt harm much to open them as root.
I think it might be the case ubuntu has these under some 'input' group other software expects to be able to use, which it cant when its opened by something running as root.

to confirm this is that:
- post the output of `ls -l /dev/input/by-id`
- try adding yourself to the input group, or whatever non root thing those are owned by on your system
- relog
- start mergeinput the same way without sudo

alternatively if its not just a non root group but also a non root user you could use sudo -u _thisuser_ with merginputs

**Comment by Ckath** [comment]
> I'm planning on rolling out debian with xfce for some people in the near future, I'll try to test it there to see if I find any issues similar to this

I ended up going with a rolling distro for those people as someone convinced me suse tumbleweed is stable enough these days. it doesnt have the issue with the xfce flavor

**Comment by gcalus** [comment]
Unfortunately it's not that, all inputs are owned by root:

$ ls -l /dev/input/by-id
total 0
lrwxrwxrwx 1 root root 11 wrz 27 14:09 usb-0911_2188-event-if01 -> ../event257
lrwxrwxrwx 1 root root 10 wrz 27 14:09 usb-0911_2188-event-kbd -> ../event31
lrwxrwxrwx 1 root root 11 wrz 27 14:09 usb-0911_2188-if01-event-mouse -> ../event256
lrwxrwxrwx 1 root root  9 wrz 27 14:09 usb-0911_2188-if01-mouse -> ../mouse4
lrwxrwxrwx 1 root root  9 wrz 27 13:32 usb-1ea7_2.4G_Mouse-event-if00 -> ../event9
lrwxrwxrwx 1 root root 10 wrz 27 13:32 usb-1ea7_2.4G_Mouse-event-mouse -> ../event10
lrwxrwxrwx 1 root root  9 wrz 27 13:32 usb-1ea7_2.4G_Mouse-mouse -> ../mouse3
lrwxrwxrwx 1 root root 10 wrz 27 13:32 usb-CN0V976R8LG00145DY9KA01_Integrated_Webcam_HD_200901010001-event-if00 -> ../event30
lrwxrwxrwx 1 root root 11 wrz 27 14:09 usb-Hantick_USB_Keyboard_HID_Composite_device-event-if01 -> ../event263
lrwxrwxrwx 1 root root 11 wrz 27 14:09 usb-Hantick_USB_Keyboard_HID_Composite_device-event-kbd -> ../event259
lrwxrwxrwx 1 root root 11 wrz 27 14:09 usb-Hantick_USB_Keyboard_HID_Composite_device-if01-event-mouse -> ../event260
lrwxrwxrwx 1 root root  9 wrz 27 14:09 usb-Hantick_USB_Keyboard_HID_Composite_device-if01-mouse -> ../mouse5
lrwxrwxrwx 1 root root 10 wrz 27 13:32 usb-Logitech_Logitech_Wireless_Headset-event-if03 -> ../event12

**Comment by Ckath** [comment]
well damn, that wouldve worked out explainably

**Comment by fneirotti** [comment]
I'm having a similar issue to https://github.com/Ckath/mergeinputs/issues/3#issuecomment-1250375536 on KDE Wayland (Arch)

Whenever I run mergeinputs with one input device (or more) kwin throws an error:

```
oct 07 22:20:00 laptop kernel: input: merged input as /devices/virtual/input/input64
oct 07 22:20:00 laptop systemd-logind[1065]: Watching system buttons on /dev/input/event18 (merged input)
oct 07 22:20:00 laptop kwin_wayland[1429]: kwin_core: Failed to open /dev/input/event18 device (No such device)
```


But I think I found something that may fix it, or at least get us closer to a solution.

ydotool makes use of `uinput` in a similar way to this tool. The only difference is the way in which the [keycodes](https://github.com/ReimuNotMoe/ydotool/blob/1d5d227e3a6dc9b70147960e99f17695d9f6804e/Daemon/ydotoold.c#L87) are set with ioctl.
It uses a predefined list taken in part from `/usr/include/linux/input-event-codes.h`

Looking into input-event-codes.h it can be seen that not all codes are defined, so the way that it iterates [sequentially](https://github.com/Ckath/mergeinputs/blob/2958a0f401b49969993ada2517443102ffaabe8b/mergeinputs.c#L69) over all possible key codes could be the culprit.

I've tried it locally myself and it works.

**Comment by Ckath** [comment]
looking at what is defined in `/usr/include/linux/input-event-codes.h` I wonder if even going with a predefined list would solve it for everyone, what if its different for another distribution and it goes over an undefined keycode again?
what if instead its an issue with it for example starting at 0, which seems `KEY_RESERVED` or some other very specific key which shouldnt be set.
also the error is really 'Failed to open device' with as backend returned error 'permission denied', it would seem very odd to me that just defining a keycode that doesnt exist and is never send creates this behaviour for uinput devices. but ofcourse I'm no expert.

that said, I have created [a branch ](https://github.com/Ckath/mergeinputs/tree/dm_issues)related to this issue which now at least has this changed to be the same as ydotool, give it a spin if you feel like it. and cheers for digging into it more

**Comment by adrixo** [comment]
Your branch solution is working for me! on Ubuntu 22.04 using x11

**Comment by Ckath** [comment]
despite being pretty sure its going to screw me over at least somewhere when I use a weird key thats not in this list, it seems to be the easiest solution for now so I'm probably just going to merge it in and see if the issue comes back
if any of the original people with the issue can confirm it does work I'll close it

**Comment by fakrep** [comment]
what a journey :o 
The workaround fixed it on my Ubuntu 22.04 with X11 too :) 

@Ckath Thank you so much!!
